File: scala/robot-simulator/src/main/scala/Robot.scala
object Bearing extends Enumeration {
  type Bearing = Value
  val North, East, South, West = Value
}

case class Robot(bearing: Bearing.Value, coordinates: (Int, Int)) {

  def turnRight(): Robot = {
    val newBearing =
      if (this.bearing == Bearing.West) {
        Bearing.North
      } else {
        Bearing(this.bearing.id + 1)
      }
    this.copy(bearing = newBearing)
  }

  def turnLeft(): Robot = {
    val newBearing =
      if (this.bearing == Bearing.North) {
        Bearing.West
      } else {
        Bearing(this.bearing.id - 1)
      }
    this.copy(bearing = newBearing)
  }

  def advance(): Robot = {
    this.bearing match {
      case Bearing.North => this.copy(coordinates = (this.coordinates._1, this.coordinates._2 + 1))
      case Bearing.West => this.copy(coordinates = (this.coordinates._1 - 1, this.coordinates._2))
      case Bearing.East => this.copy(coordinates = (this.coordinates._1 + 1, this.coordinates._2))
      case Bearing.South => this.copy(coordinates = (this.coordinates._1, this.coordinates._2 - 1))
    }
  }

  def simulate(seq: String): Robot = {
    seq.foldLeft(this)({
      case (robot, instruct) =>
        instruct match {
          case 'R' => robot.turnRight()
          case 'L' => robot.turnLeft()
          case 'A' => robot.advance()
        }
    })
  }
}

File: haskell/robot-simulator/src/Robot.hs
module Robot
    ( Bearing(East,North,South,West)
    , bearing
    , coordinates
    , mkRobot
    , move
    ) where

data Bearing = North
             | East
             | South
             | West
             deriving (Eq, Show, Enum, Bounded)

data Robot = Robot Bearing Integer Integer

bearing :: Robot -> Bearing
bearing (Robot b _ _) = b

coordinates :: Robot -> (Integer, Integer)
coordinates (Robot _ x y) = (x, y)

mkRobot :: Bearing -> (Integer, Integer) -> Robot
mkRobot direction coordinates = Robot direction (fst coordinates) (snd coordinates)

move :: Robot -> String -> Robot
move robot instructions = foldl instruct robot instructions
    where
        instruct r i
            | i == 'L' = mkRobot (if robotBearing == minBound then maxBound else pred robotBearing) robotCoords
            | i == 'R' = mkRobot (if robotBearing == maxBound then minBound else succ robotBearing) robotCoords
            | i == 'A' = mkRobot robotBearing (fst robotCoords + fst robotAdvanceCoords, snd robotCoords + snd robotAdvanceCoords)
            | otherwise = error "panic"
            where
                robotBearing = bearing r
                robotCoords  = coordinates r
                getAdvanceCoords North = (0, 1)
                getAdvanceCoords South = (0, -1)
                getAdvanceCoords East  = (1, 0)
                getAdvanceCoords West  = (-1, 0)
                robotAdvanceCoords = getAdvanceCoords robotBearing